### PR TITLE
[webview_flutter] Add interface for showing javascript dialog message

### PR DESCRIPTION
**Pull request by jsharp83**
* there are cases where Web calls System Popup with javascript on webview_flutter
* At this time, the message comes in the WKUIDelegate part in iOS.
   * https://developer.apple.com/documentation/webkit/wkuidelegate/1537406-webview
   * https://developer.apple.com/documentation/webkit/wkuidelegate/1536489-webview
* Android also has a interface on WebChromeClient
   * https://developer.android.com/reference/android/webkit/WebChromeClient#onJsAlert(android.webkit.WebView,%20java.lang.String,%20java.lang.String,%20android.webkit.JsResult)
* It was implemented according to the requirements of the code review of the https://github.com/flutter/packages/pull/4555
* Related issue: https://github.com/flutter/flutter/issues/30358#issuecomment-1645347616
* Related Interface PR: https://github.com/flutter/packages/pull/5670

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by jsharp83** [comment]
@bparrishMines Hi, Can you check this PR? This is my first contribution here so I don't know exact routine for contribution. Help me please. Thanks :)

File: packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md:4]: nit:
```suggestion
* Adds support to show JavaScript dialog. See `PlatformWebViewController.setJavaScriptAlertDialogCallback`.
```

This can also list the other two callback methods.

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart:285]: Since this is only setting one callback method, the pattern should be 
```suggestion
  Future<void> setOnJavaScriptAlertDialog(
      Future<void> Function(String message)
          onJavaScriptAlertDialog) async {
```

Same for the two methods below.

File: packages/webview_flutter/webview_flutter_platform_interface/pubspec.yaml
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/pubspec.yaml:7]: This will be a minor version bump since it is adding a new feature. (e.g. `2.6.0`)

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebChromeClientHostApiImpl.java
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebChromeClientHostApiImpl.java:43]: Since the platform interface separates handling the callbacks into three separate methods, you should also have three separate flags for each `alert`/`confirm`/`prompt`. What happens if a user only wants to set the `confirm` dialog?

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:765]: Since the native `WKUIDelegate` separates handling dialogs into separate methods, this should also.

**Review by bparrishMines — COMMENTED** [review]
Thanks for implementing this for Android and iOS! I left my first comments. 

Also, we're working on updating `pigeon` to handle more of the wrapping code if you would prefer to wait for that before continuing.

**Comment by jsharp83** [comment]
@bparrishMines It has been modified according to the review. Could you please check the code again?

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:1173]: This and the two methods below should follow the same pattern as `onShowFileChooser` and be called `setSynchronousReturnValueForOnJsAlert`.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:1136]: These should have the same name as the corresponding Java method. (e.g. `onJsAlert`). They should also have a return value of void since they can't return a synchronous value.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:1142]: These can be removed. I don't believe they are used and they are redundant with the callback methods above.

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:744]: This should be `runJavaScriptAlertPanel`. The `withIdentifier` should only be added to the `@ObjCSelector` annotation above and not in the method.

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:221]: This needs to be `weakThis.target?._onJavaScriptAlertDialogCallback`. The methods below also need to be updated.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart
Comment (jsharp83) [reply packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart]: Sorry, I cannot fully understand on your comment.

> They should also have a return value of void since they can't return a synchronous value.

They should have a Future value to pass the value for JSResult and it works well in my test.

Simulator Screen Recording - iPhone 13 mini - 2023-09-28 at 15 04 27

FileChooser in android_webview also return a List<String> previously, So I just follow it to implement.

Can you explain more to fix this?
I really appreciate your help :)

Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart]: Ah, now I understand. Your implementation is correct. I didn't notice the `JSResult` in the params.

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart:293]: It looks like [WebChromeClient.onJsAlert](https://developer.android.com/reference/android/webkit/WebChromeClient#onJsAlert(android.webkit.WebView,%20java.lang.String,%20java.lang.String,%20android.webkit.JsResult)) and [WKUIDelegate.runJavaScriptAlertPanelWithMessage](https://developer.apple.com/documentation/webkit/wkuidelegate/1537406-webview?language=objc) also provide the URL for these callbacks. We should change `String message` to a class that contains a `message` field where we can add additional information at a later point. You could create a class named something like `JavaScriptAlertDialogRequest`. See [JavaScriptConsoleMessage](https://github.com/flutter/packages/blob/main/packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/javascript_console_message.dart) as an example.

This goes for the methods below as well. (e.g. `JavaScriptConfirmDialogRequest` and `JavaScriptTextInputDialogRequest`).

File: packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md:3]: nit:

```suggestion
* Adds support to show JavaScript dialog. See
  `PlatformWebViewController.setOnJavaScriptAlertDialog`,
  `PlatformWebViewController.setOnJavaScriptConfirmDialog`, and
  `PlatformWebViewController.setOnJavaScriptTextInputDialog`.
```

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart
Comment (jsharp83) [reply packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart]: I fixed it. Can you check it again?

**Comment by jsharp83** [comment]
@bparrishMines Can you check if there are any more areas that need to be revised?

**Comment by rekire** [comment]
What is the current status here? It looks good so far out of my view

**Comment by jsharp83** [comment]
@bparrishMines Hi, Can you check this please? I would like to merge this PR and move on to other issues. Thanks

File: packages/webview_flutter/webview_flutter_android/android/gradle/wrapper/gradle-wrapper.properties
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/android/gradle/wrapper/gradle-wrapper.properties:1]: You shouldn't need to add this file

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:1142]: I may have missed these before, but this callback method should have the same method signature as the Java method: https://developer.android.com/reference/android/webkit/WebChromeClient#onJsAlert(android.webkit.WebView,%20java.lang.String,%20java.lang.String,%20android.webkit.JsResult).  As in `(WebView, String,  String)`. 

The callback methods below should also follow the method signature of their Java methods as well.

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart:795]: These also need to follow the method signature of the corresponding Objc class: https://developer.apple.com/documentation/webkit/wkuidelegate/1537406-webview?language=objc.

Comment (jsharp83) [reply packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart]: @bparrishMines 

I appreciate your review.
To follow the method signature, I needed to add request field on WKFrameInfo.
Can you check last commit?
Thanks :)

**Comment by rekire** [comment]
I tried to check your changes locally so I changed my dependencies like this:
```yaml
dependencies:
  # other dependencies
  webview_flutter: #^4.2.2
    git:
      url: https://github.com/jsharp83/packages.git
      ref: 38506f2c831db7a64ab593ae33de5fe73ef44a24
      path: packages/webview_flutter/webview_flutter

dependency_overrides:
  webview_flutter_android:
    git:
      url: https://github.com/jsharp83/packages.git
      ref: 38506f2c831db7a64ab593ae33de5fe73ef44a24
      path: packages/webview_flutter/webview_flutter_android
  webview_flutter_platform_interface:
    git:
      url: https://github.com/jsharp83/packages.git
      ref: 38506f2c831db7a64ab593ae33de5fe73ef44a24
      path: packages/webview_flutter/webview_flutter_platform_interface
  webview_flutter_wkwebview:
    git:
      url: https://github.com/jsharp83/packages.git
      ref: 38506f2c831db7a64ab593ae33de5fe73ef44a24
      path: packages/webview_flutter/webview_flutter_wkwebview
```

After syncing I noticed that the `WebViewController` does how contain your new methods `setOnJavaScriptAlertDialog`, `setOnJavaScriptConfirmDialog` and `setOnJavaScriptTextInputDialog`.

Can you tell me what I missed?

**Comment by jsharp83** [comment]
@rekire 

Can you add that methods like this?

https://github.com/flutter/packages/pull/4704/files#diff-9d105bffc5821b3c24f256092b794212ad8d8bccf547b59061f62334f5bf2a80R527-R547

If it is not work, I will test a code on new repo. Thanks :)

**Comment by rekire** [comment]
I tried that but I get this error:
> The method 'setOnJavaScriptAlertDialog' isn't defined for the class 'WebViewController'.

I cannot see that referenced methods in https://github.com/jsharp83/packages/blob/add-javascript-panel-interface/packages/webview_flutter/webview_flutter/lib/src/webview_controller.dart they should be there isn't it?

**Comment by rekire** [comment]
Out of my view you forgot to expose your new interface. Check my commit https://github.com/rekire/flutter-packages-fork/commit/aa23cec5bbc0b4e953453435fb42d85c9e1524aa. Feel free to cherry pick it. Or if you prefer I could make a PR for that.

**Comment by rekire** [comment]
That was quick :-D

**Comment by jsharp83** [comment]
@rekire Thanks for your help :) I already pushed the commit. so I'm going to fix my repo to cherry pick your commit.

**Comment by rekire** [comment]
That's great I could add with your help now the confirm dialog. I can use for now your fork to add this missing feature until it became merged.

For other early adapters you can try (but I do not suggest):
```yaml
dependencies:
  webview_flutter:
    # Remove this git dependency when https://github.com/flutter/packages/pull/4704 is merged and published
    git:
      url: https://github.com/jsharp83/packages.git
      ref: add-javascript-panel-interface
      path: packages/webview_flutter/webview_flutter

# Remove this when https://github.com/flutter/packages/pull/4704 is merged and published
dependency_overrides:
  webview_flutter_android:
    git:
      url: https://github.com/jsharp83/packages.git
      ref: add-javascript-panel-interface
      path: packages/webview_flutter/webview_flutter_android
  webview_flutter_platform_interface:
    git:
      url: https://github.com/jsharp83/packages.git
      ref: add-javascript-panel-interface
      path: packages/webview_flutter/webview_flutter_platform_interface
  webview_flutter_wkwebview:
    git:
      url: https://github.com/jsharp83/packages.git
      ref: add-javascript-panel-interface
      path: packages/webview_flutter/webview_flutter_wkwebview
```

For the WebViewController you need this:
```dart
final webViewController = WebViewController()
  ..setOnJavaScriptAlertDialog((JavaScriptAlertDialogRequest request) async {
    await _showAlert(context, request.message);
  })
  ..setOnJavaScriptConfirmDialog((JavaScriptConfirmDialogRequest request) async {
    return _showConfirm(context, request.message);
  })
  ..setOnJavaScriptTextInputDialog((JavaScriptTextInputDialogRequest request) async {
    return _showTextInput(context, request.message, request.defaultText);
  });
```

And the dialogs from the example app:
```dart
  Future<void> _showAlert(BuildContext context, String message) async {
    return showDialog(
        context: context,
        builder: (BuildContext ctx) {
          return AlertDialog(
            content: Text(message),
            actions: <Widget>[
              TextButton(
                  onPressed: () {
                    Navigator.of(ctx).pop();
                  },
                  child: const Text('OK'))
            ],
          );
        });
  }

  Future<bool> _showConfirm(BuildContext context, String message) async {
    final dynamic result = await showDialog(
        context: context,
        builder: (BuildContext ctx) {
          return AlertDialog(
            content: Text(message),
            actions: <Widget>[
              TextButton(
                  onPressed: () {
                    Navigator.of(ctx).pop(false);
                  },
                  child: const Text('Cancel')),
              TextButton(
                  onPressed: () {
                    Navigator.of(ctx).pop(true);
                  },
                  child: const Text('OK')),
            ],
          );
        });

    return result.runtimeType == bool && result as bool;
  }

  Future<String> _showTextInput(BuildContext context, String message, String? defaultText) async {
    final dynamic result = await showDialog(
        context: context,
        builder: (BuildContext ctx) {
          return AlertDialog(
            content: Text(message),
            actions: <Widget>[
              TextButton(
                  onPressed: () {
                    Navigator.of(ctx).pop('Text test');
                  },
                  child: const Text('Enter')),
            ],
          );
        });

    return result.runtimeType == String ? result as String : '';
  }
```

**Comment by jsharp83** [comment]
@bparrishMines I know you've seen this multiple times and it might be annoying, but I would appreciate a review. This is an old PR and I would like to merge it quickly to move on to other tasks. Thank you.

**Comment by rekire** [comment]
To be honest I have no idea how to make the build green, but that might be a good task to find out first.

**Comment by stuartmorgan** [comment]
> * [x]  All existing and new tests are passing.

Please see https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins

This PR hasn't followed step 1 there, which means most of the CI checks can't run correctly.

**Comment by stuartmorgan** [comment]
Also:
 
> * [x]  I added new tests to check the change I am making, or this PR is [test-exempt](https://github.com/flutter/flutter/wiki/Tree-hygiene#tests).

I see tests that the platform interfaces defaults to unimplemented, but I don't see any tests of the actual new functionality in any of the other packages.

**Comment by stuartmorgan** [comment]
> But it still fails in some check of CI. Is it OK?

Only the `federated safety check` and `publishability` failures are expected; the rest appear to be real issues in the PR.

**Comment by jsharp83** [comment]
@bparrishMines @stuartmorgan 

Hi, I've fixed the CI issue, and now there is only one failure remaining.
> Tasks failed: federated safety check

Please check this PR

File: packages/webview_flutter/webview_flutter_wkwebview/CHANGELOG.md
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/CHANGELOG.md:30]: nit:

```suggestion
```

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart:795]: nit:

```suggestion
  /// Notifies the host application that the web page
```

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart:800]: nit:

```suggestion
  /// Notifies the host application that the web page
```

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart:805]: nit:

```suggestion
  /// Notifies the host application that the web page
```

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFUIDelegateHostApi.m
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFUIDelegateHostApi.m:117]: Since `NSAssert` is ignored in production, we should change this to

```suggestion
                                                  if (error) {
                                                    completionHandler(NO);
                                                  } else {
                                                    completionHandler(isConfirmed.boolValue);
                                                  }
```

We have an issue with another method that doesn't ensure this is always called.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFUIDelegateHostApi.m
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFUIDelegateHostApi.m:136]: Same for this

```suggestion
                                                  if (error) {
                                                    completionHandler(nil);
                                                  } else {
                                                    completionHandler(inputText);
                                                  }  
```

**Review by bparrishMines — COMMENTED** [review]
This mostly looks good to me. I only have a few more comments and the `webview_flutter` package is missing tests. 

@stuartmorgan I think this is ready for a secondary review.

**Comment by jsharp83** [comment]
> This mostly looks good to me. I only have a few more comments and the `webview_flutter` package is missing tests.

@bparrishMines I added test cases on webview_flutter. Thank you :)

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.m
Comment (hellohuanlin) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.m:170]: only `@""` can have 0 length right

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFGeneratedWebKitApis.h
Comment (hellohuanlin) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFGeneratedWebKitApis.h:881]: what's the diff among the 3 options? do you have some screenshot?

**Comment by jsharp83** [comment]
> This mostly looks good to me. I only have a few more comments and the `webview_flutter` package is missing tests.
> 
> @stuartmorgan I think this is ready for a secondary review.

@stuartmorgan Can I make the platform package interface PR for secondary review? Thanks

**Review by stuartmorgan — COMMENTED** [review]
Overall this looks good to me; please feel free to make the platform interface PR!

**Comment by jsharp83** [comment]
@bparrishMines @stuartmorgan @hellohuanlin Can you check this again? The platfrom interface PR has been merged. Thanks :)

**Comment by stuartmorgan** [comment]
> @bparrishMines @stuartmorgan @hellohuanlin Can you check this again? The platform interface PR has been merged. Thanks :)

The next step is to split out a landable PR for the implementation packages, then we'll do final review on that.

**Comment by jsharp83** [comment]
@bparrishMines @stuartmorgan @hellohuanlin 

HI, This PR only contains webview_flutter package part now.
I needed to update minimum flutter version on webview_flutter package due to webview_flutter_wkwebview
Please check this again. Thank you.

File: packages/webview_flutter/webview_flutter/CHANGELOG.md
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/CHANGELOG.md:3]: Nit: "Adds support for custom handling of JavaScript dialogs. See [...]"

Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter/CHANGELOG.md:3]: (Also, please wrap to ~80 characters.)

**Review by stuartmorgan — COMMENTED** [review]
LGTM with CHANGELOG nit. @bparrishMines for secondary approval.

**Comment by rekire** [comment]
@jsharp83 I think you missed that line break change request. I cannot await that this goes finally live.

File: packages/webview_flutter/webview_flutter/CHANGELOG.md
Comment (rekire) [inline packages/webview_flutter/webview_flutter/CHANGELOG.md:4]: I think this line is still too long

File: packages/webview_flutter/webview_flutter/CHANGELOG.md
Comment (rekire) [reply packages/webview_flutter/webview_flutter/CHANGELOG.md:10]: This two lines can be removed you have the content now in 4.6.0

**Review by bparrishMines — APPROVED** [review]
LGTM

**Comment by bparrishMines** [comment]
@stuartmorgan It looks like this needs the approval along with your LGTM

**Review by stuartmorgan — APPROVED** [review]
LGTM; looks like I just pressed the wrong button last time.

**Pull request merged**
